pkgpanda/build/__init__.py

```diff
@@ -616,20 +606,20 @@ def visit(pkg_tuple):
     def key_func(elem):
         return elem[0], elem[1] is None, elem[1] or ""
 
-    def visit_packages_in_tree(variant):
-        all_tuples = get_tree_package_tuples(package_store, variant)
-        for pkg_tuple in sorted(all_tuples, key=key_func):
+    def visit_packages(package_tuples):
+        for pkg_tuple in sorted(package_tuples, key=key_func):
             if pkg_tuple in visited:
                 continue
             visit(pkg_tuple)
 
-    if tree_variant is None:
-        tree_visit_list = list(sorted(package_store.list_trees(), key=pkgpanda.util.variant_str))
+    if tree_variant:
+        package_sets = [package_store.get_package_set(tree_variant)]
     else:
-        tree_visit_list = [tree_variant]
+        package_sets = package_store.get_all_package_sets()
 
-    for tree in tree_visit_list:
-        visit_packages_in_tree(tree)
+    # Build all required packages for all tree variants.
```

Comment:
I pushed a comment fixup commit that documents that detail in `visit()`: https://github.com/dcos/dcos/pull/300/commits/4fa96a505fc9a272cd246352984297e5038075a7